### PR TITLE
Merge several version of MsQuicStream SendAsync code

### LINKED ISSUE #55818: [QUIC] Merge several version of stream SendAsync
Why do we have 3 version of sending? We should unite as much as possible.
Are we even testing all of them?
One of them is for Kestrel - `SendReadOnlySequenceAsync`

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #55818.

This PR extracts common code from several `Send*Async` methods used to implement `WriteAsync` into a single `WriteAsyncInternal` method.

### CODE REVIEW

**Comment by rzikm** [comment]
Test failures seem to be all some permission errors, no actual test failures as far as I can see

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1718]: I haven't looked much into this yet, but it feels like it might be better to wait for https://github.com/dotnet/runtime/pull/68288 to go in to be able to reuse stuff like `MsQuicBuffers` that was added there?

**Comment by rzikm** [comment]
Test failure is unrelated (`System.Net.Primitives.UnitTests.Tests`)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1718]: Yes, please wait with this on my PR.

**Comment by rzikm** [comment]
Putting this on hold until https://github.com/dotnet/runtime/pull/68288 is in.

**Comment by rzikm** [comment]
I reused the `MsQuicBuffers` struct @ManickaP added, but I think I still need to keep the adaptor structs, unless we are fine with replacing it with lambdas. I briefly tried managed function pointers, but they can't be used in non-unsafe context, and you can't await in unsafe context....

**Comment by rzikm** [comment]
Since #68288 got reverted in #68940, we need to put this PR on hold

**Comment by rzikm** [comment]
Test failures are unrelated:
- System.Text.RegularExpressions.Tests.RegexIgnoreCaseTests.Unicode_IgnoreCase_Tests
- LibraryImportGenerator.UnitTests.Compiles.ValidateSnippetsFallbackForwarder

**Review by ManickaP — COMMENTED** [review]
In general looks good, thanks.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs:44]: The size of `_buffers` is always the same as `_handles`, so they will either both grow or neither of them will.

`_count` contains the number of valid slots in those arrays, depending on history, it may be lower than `_handles.Length`

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1705]: I was influenced by how `SslStream` internals are implemented.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Security/src/System/Net/Security/ReadWriteAdapter.cs

It possibly does not make as much sense to do it here as well. The upside of using these is that the compiler can inline the calls (which it cannot do for delegates/lambdas). If we don't care about inlining/virtual dispatch for each call, I can change it to use delegates.

Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1705]: Ok, I'll answer this with more questions :smile: 
- Do you think that the inlining/non-inlining is something that'd make a dent in perf? Or is rather negligible in the grand scheme of things (aka send itself)?
- Which one do you think is easier to read, understand and maintain: interface or delegate?

I don't have an answer to the first question (we could measure the difference with microbenchmarks), and for the second I have only my subjective opinion.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1705]: Right, the perf impact is probably going to be negligible. We can always come back to it later.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:311]: One last tiny nit:
```suggestion
        private unsafe ValueTask WriteAsyncCore<TBuffer>(Action<State, TBuffer> stateSetup, TBuffer buffer, bool isEmpty, bool endStream)
```

I think you can push the flags creation even one layer down. `if ((flags & QUIC_SEND_FLAGS.FIN) == QUIC_SEND_FLAGS.FIN)` is basically `if (endStream)`.

**Review by ManickaP — APPROVED** [review]
Looks good, thanks!

**Review by CarnaViire — APPROVED** [review]
LGTM modulo nits, thanks!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:86]: Putting `= new MsQuicBuffers()` in the field initializer triggers "Don't assign default values" or similar warning, and leaving it uninitialized breaks because the `MsQuicBuffers._handles` is null.

**Pull request merged**
